File: app/src/main/java/org/deiverbum/app/data/entity/relation/TodayMisaLecturas.java
package org.deiverbum.app.data.entity.relation;

import androidx.room.Embedded;
import androidx.room.Relation;

import org.deiverbum.app.data.entity.LiturgyEntity;
import org.deiverbum.app.data.entity.MassReadingEntity;
import org.deiverbum.app.data.entity.TodayEntity;
import org.deiverbum.app.model.MassReading;
import org.deiverbum.app.model.MassReadingList;
import org.deiverbum.app.model.Today;

import java.util.ArrayList;
import java.util.List;

/**
 * @author A. Cedano
 * @version 1.0
 * @since 2023.1
 */
public class TodayMisaLecturas {

    @Embedded
    public TodayEntity today;

    @Relation(
            entity = LiturgyEntity.class,
            parentColumn = "massReadingFK",
            entityColumn = "liturgyID"
    )
    public LiturgyWithTime feria;

    @Relation(
            entity = LiturgyEntity.class,
            parentColumn = "previousFK",
            entityColumn = "liturgyID"
    )
    public LiturgyWithTime previo;

    @Relation(
            entity = MassReadingEntity.class,
            parentColumn = "massReadingFK",
            entityColumn = "liturgyFK")
    public List<MassReadingWithAll> lecturas;

    public Today getToday(){
        Today dm = new Today();
        dm.liturgyDay=feria.getDomainModel();
        dm.liturgyPrevious=today.previoId>1?previo.getDomainModel():null;
        dm.setTodayDate(today.getHoy());
        dm.setHasSaint(today.hasSaint);
        dm.setMLecturasFK(today.mLecturasFK);
        return dm;
    }

    public MassReadingList getDomainModel(){
        MassReadingList dm=new MassReadingList();
        dm.setToday(getToday());
        List<MassReading> listModel = new ArrayList<>();
        for (MassReadingWithAll item : lecturas) {
                listModel.add(item.getDomainModel());
        }
        dm.setLecturas(listModel);
        dm.sort();
        return dm;
    }

}

File: app/src/main/java/org/deiverbum/app/data/entity/relation/TodayMisaLecturas.kt
package org.deiverbum.app.data.entity.relation

import androidx.room.Embedded
import androidx.room.Relation
import org.deiverbum.app.data.entity.LiturgyEntity
import org.deiverbum.app.data.entity.MassReadingEntity
import org.deiverbum.app.data.entity.MassReadingJoinEntity
import org.deiverbum.app.data.entity.TodayEntity
import org.deiverbum.app.model.MassReading
import org.deiverbum.app.model.MassReadingList
import org.deiverbum.app.model.Today

/**
 * @author A. Cedano
 * @version 1.0
 * @since 2023.1
 */
data class TodayMisaLecturas(
    @Embedded
    var today: TodayEntity,

    @Relation(
        entity = LiturgyEntity::class,
        parentColumn = "massReadingFK",
        entityColumn = "liturgyID"
    )
    var feria: LiturgyWithTime,

    @Relation(
        entity = LiturgyEntity::class,
        parentColumn = "previousFK",
        entityColumn = "liturgyID"
    )
    var previo: LiturgyWithTime,

    @Relation(
        entity = MassReadingJoinEntity::class,
        parentColumn = "massReadingFK",
        entityColumn = "liturgyFK"
    )
    var joinTable: MassReadingJoinEntity,

    @Relation(
        entity = MassReadingEntity::class,
        parentColumn = "massReadingFK",
        entityColumn = "liturgyFK"
    )
    var lecturas: List<MassReadingWithAll>
) {
    private fun getToday(): Today {
        val dm = Today()

        dm.liturgyDay = feria.domainModel
        dm.liturgyPrevious = if (today.previousFK > 1) previo.domainModel else null
        dm.todayDate = today.todayDate
        dm.hasSaint = today.hasSaint
        //dm.todayDate = hoy

        dm.setMLecturasFK(today.massReadingFK)
        return dm
    }

    val domainModel: Today
        get() {
            val dmToday = getToday()
            val dm = feria.domainModel
            dm.typeID = 10

            val mr = MassReadingList()
            mr.type = joinTable.type
            val listModel: MutableList<MassReading?> = ArrayList()
            for (item in lecturas) {
                listModel.add(item.domainModel)
            }
            listModel.sortBy { it?.getOrden() }
            mr.lecturas = listModel
            mr.today=getToday()
            //mr.sort()
            dmToday.liturgyDay = dm
            dmToday.liturgyDay.massReadingList = mr
            return dmToday
        }
}